### PR TITLE
net: sntp: support building without NET_SOCKETS_POSIX_NAMES

### LINKED ISSUE #57133: Compile error building SNTP library for native_posix while NET_SOCKETS_POSIX_NAMES=n
**Describe the bug**
It's not possible to build an application with the SNTP library using the `native_posix` board when `CONFIG_NET_SOCKETS_POSIX_NAMES=n` and `CONFIG_POSIX_API=n`. The error shown is

```
In file included from /home/hein/zephyrproject/zephyr/include/zephyr/posix/unistd.h:10,
                 from /home/hein/zephyrproject/zephyr/include/zephyr/net/sntp.h:15,
                 from /home/hein/zephyrproject/zephyr/samples/net/sockets/sntp_client/src/main.c:11:
/home/hein/zephyrproject/zephyr/include/zephyr/posix/sys/stat.h:40:10: fatal error: sys/_timespec.h: No such file or directory
   40 | #include <sys/_timespec.h>
      |          ^~~~~~~~~~~~~~~~~
```
A previous PR #50805 was merged by a member of our team to allow building SNTP for the same scenario, but I'm relatively confident that this functionality was broken again with a necessary change in https://github.com/zephyrproject-rtos/zephyr/commit/6f4e96bc2482e868590f3f72cfed92d14dcd6de1. 

After conferring with @cfriedt he mentioned that this could be fixed by changing the headers in `sntp.h`, something like:
```
#ifdef CONFIG_ARCH_POSIX
#include <sys/socket.h>
#include <unistd.h>
#include <poll.h>
#else
#include <zephyr/posix/sys/socket.h>
#include <zephyr/posix/unistd.h>
#include <zephyr/posix/poll.h>
#endif
```

I could get it to work, but it turned into a little oil spill where it required fixing a few files named below. I'm not sure what is the right way to fix it, and the fix seemed quite messy. If no one else is willing I will create a draft PR with these fixes soon.
```
include/zephyr/net/sntp.h
include/zephyr/net/socketutils.h
subsys/net/lib/sntp/sntp.c
subsys/net/lib/sntp/sntp_simple.c
subsys/net/lib/utils/addr_utils.c
```

**To Reproduce**
1. In `zephyr/samples/net/sockets/sntp_client/prj.conf` change to `CONFIG_NET_SOCKETS_POSIX_NAMES=n`.
2. `west build -p always -b native_posix samples/net/sockets/sntp_client`
4. Witness the mentioned compiler error

**Expected behavior**
For the application to build without warnings.

**Impact**
This is currently a showstopper in our attempt to migrate from Zephyr 2.6 to the upcoming Zephyr 3.4.

**Environment (please complete the following information):**
 - Linux
 - Zephyr SDK 0.16
 - Zephyr commit: 3707014f20f7c9fcb71e4f13fd23cf897b1dcbdc

### PR DESCRIPTION
**Pull request by heinwessels**
Fixes the compiler error that occurs when attempting to build the SNTP library with CONFIG_NET_SOCKETS_POSIX_NAMES=n, while using native_posix.

Closes #57133

### CODE REVIEW

File: subsys/net/lib/sntp/sntp.c [hunk 1]
```diff
@@ -106,7 +107,7 @@ static int sntp_recv_response(struct sntp_ctx *sntp, uint32_t timeout,
 	int status;
 	int rcvd;
 
-	status = zsock_poll(sntp->sock.fds, sntp->sock.nfds, timeout);
```
Comment (cfriedt) [inline subsys/net/lib/sntp/sntp.c:109]: this should probably be unchanged

File: subsys/net/lib/sntp/sntp.c [hunk 2]
```diff
@@ -116,7 +117,7 @@ static int sntp_recv_response(struct sntp_ctx *sntp, uint32_t timeout,
 		return -ETIMEDOUT;
 	}
 
-	rcvd = zsock_recv(sntp->sock.fd, (uint8_t *)&buf, sizeof(buf), 0);
+	rcvd = recv(sntp->sock.fd, (uint8_t *)&buf, sizeof(buf), 0);
```
Comment (cfriedt) [inline subsys/net/lib/sntp/sntp.c:120]: this should probably be unchanged

File: subsys/net/lib/sntp/sntp.c [hunk 3]
```diff
@@ -150,15 +151,15 @@ int sntp_init(struct sntp_ctx *ctx, struct sockaddr *addr, socklen_t addr_len)
 
 	memset(ctx, 0, sizeof(struct sntp_ctx));
 
-	ctx->sock.fd = zsock_socket(addr->sa_family, SOCK_DGRAM, IPPROTO_UDP);
+	ctx->sock.fd = socket(addr->sa_family, SOCK_DGRAM, IPPROTO_UDP);
```
Comment (cfriedt) [inline subsys/net/lib/sntp/sntp.c:154]: this should probably be unchanged

File: subsys/net/lib/sntp/sntp.c [hunk 4]
```diff
@@ -150,15 +151,15 @@ int sntp_init(struct sntp_ctx *ctx, struct sockaddr *addr, socklen_t addr_len)
...
 	if (ctx->sock.fd < 0) {
 		NET_ERR("Failed to create UDP socket %d", errno);
 		return -errno;
 	}
 
-	ret = zsock_connect(ctx->sock.fd, addr, addr_len);
+	ret = connect(ctx->sock.fd, addr, addr_len);
```
Comment (cfriedt) [inline subsys/net/lib/sntp/sntp.c:160]: this should probably be unchanged

File: subsys/net/lib/sntp/sntp.c [hunk 5]
```diff
@@ -150,15 +151,15 @@ int sntp_init(struct sntp_ctx *ctx, struct sockaddr *addr, socklen_t addr_len)
...
 	if (ret < 0) {
-		(void)zsock_close(ctx->sock.fd);
+		(void)close(ctx->sock.fd);
```
Comment (cfriedt) [inline subsys/net/lib/sntp/sntp.c:162]: this should probably be unchanged

File: subsys/net/lib/sntp/sntp.c [hunk 6]
```diff
@@ -186,7 +187,7 @@ int sntp_query(struct sntp_ctx *ctx, uint32_t timeout, struct sntp_time *time)
 	ctx->expected_orig_ts = get_uptime_in_sec() + OFFSET_1970_JAN_1;
 	tx_pkt.tx_tm_s = htonl(ctx->expected_orig_ts);
 
-	ret = zsock_send(ctx->sock.fd, (uint8_t *)&tx_pkt, sizeof(tx_pkt), 0);
+	ret = send(ctx->sock.fd, (uint8_t *)&tx_pkt, sizeof(tx_pkt), 0);
```
Comment (cfriedt) [inline subsys/net/lib/sntp/sntp.c:190]: this should probably be unchanged

File: subsys/net/lib/sntp/sntp.c [hunk 7]
```diff
@@ -198,6 +199,6 @@ int sntp_query(struct sntp_ctx *ctx, uint32_t timeout, struct sntp_time *time)
 void sntp_close(struct sntp_ctx *ctx)
 {
 	if (ctx) {
-		(void)zsock_close(ctx->sock.fd);
+		(void)close(ctx->sock.fd);
```
Comment (cfriedt) [inline subsys/net/lib/sntp/sntp.c:202]: this should probably be unchanged

File: subsys/net/lib/sntp/sntp.c (hunk 2 above)
Comment (heinwessels) [reply subsys/net/lib/sntp/sntp.c:120]: It's using the `zsock_` version the prefered way?

If I try to change it back the compiler doesn't know what `zsock_recv` is for example, and they all fail:
```
/home/hein/zephyrproject/zephyr/subsys/net/lib/sntp/sntp.c:111:18: warning: implicit declaration of function ‘zsock_poll’; did you mean ‘k_poll’? [-Wimplicit-function-declaration]
  111 |         status = zsock_poll(sntp->sock.fds, sntp->sock.nfds, timeout);
      |                  ^~~~~~~~~~
      |                  k_poll
```
Which means I would have to `#include <zephyr/net/socket.h>`, but that clashes with `#include <sys/socket.h>` in `sntp.h` and creates many redeclaration errors:
```
In file included from /usr/include/sys/socket.h:26,
                 from /usr/include/netinet/in.h:23,
                 from /usr/include/arpa/inet.h:22,
                 from /home/hein/zephyrproject/zephyr/include/zephyr/net/sntp.h:12,
                 from /home/hein/zephyrproject/zephyr/subsys/net/lib/sntp/sntp.c:12:
/usr/include/bits/types/struct_iovec.h:26:8: error: redefinition of ‘struct iovec’
   26 | struct iovec
      |        ^~~~~
In file included from /home/hein/zephyrproject/zephyr/include/zephyr/net/socket.h:27,
                 from /home/hein/zephyrproject/zephyr/subsys/net/lib/sntp/sntp.c:11:
/home/hein/zephyrproject/zephyr/include/zephyr/net/net_ip.h:229:8: note: originally defined here
  229 | struct iovec {
      |        ^~~~~
```
And I I replace `#include <sys/socket.h>` with `#include <zephyr/net/socket.h>` (and remove `<poll>`) it also doesn't work, and results in a new cascade of errors:
```
In file included from /usr/include/arpa/inet.h:22,
                 from /home/hein/zephyrproject/zephyr/include/zephyr/net/sntp.h:12,
                 from /home/hein/zephyrproject/zephyr/subsys/net/lib/sntp/sntp_simple.c:9:
/home/hein/zephyrproject/zephyr/include/zephyr/net/net_ip.h:63:9: error: redeclaration of enumerator ‘IPPROTO_IP’
   63 |         IPPROTO_IP = 0,            /**< IP protocol (pseudo-val for setsockopt() */
      |         ^~~~~~~~~~
/usr/include/netinet/in.h:42:5: note: previous definition of ‘IPPROTO_IP’ with type ‘enum <anonymous>’
   42 |     IPPROTO_IP = 0,        /* Dummy protocol for TCP.  */
      |     ^~~~~~~~~~
```
I played around with changing more includes, but I always run into redeclaration errors or implicit declarations. I'm also not sure which is the right way to go about it. The only way I could get it to compile was doing this change. But please let me know if there is a better way :)

File: samples/net/sockets/sntp_client/sample.yaml [hunk 8]
```diff
@@ -3,9 +3,8 @@ sample:
   name: sntp_client
 common:
   harness: net
-  platform_allow: qemu_x86
+  platform_allow: qemu_x86 native_posix
   tags: net
-  filter: TOOLCHAIN_HAS_NEWLIB == 1
 tests:
```
Comment (rlubos) [inline samples/net/sockets/sntp_client/sample.yaml:8]: This should actually now build w/o any POSIX support being enabled at all, so maybe add a "raw" configuration as well?

**Review by cfriedt — COMMENTED** [review]
Agree with Robert's assessment. It's a bit of a layering violation to have the network stack depend on POSIX, since POSIX net APIs should wrap the zephyr ones. The opposite would introduce cyclic dependencies.

**Review by rlubos — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by cfriedt** [comment]
@heinwessels - thanks for the PR and for adding to the testsuite 👍

**Pull request merged**